### PR TITLE
specify mesos framework role

### PR DESCRIPTION
**Pull request by jduan**
This attempts to address issue 68. Please take a look.

I didn't update/add any tests but the existing tests passed.

### CODE REVIEW

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/help-role.html [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<div>
+    Specifies the role that will be used for invoking the Jenkins slaves. If
```
Comment (vinodkone) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/help-role.html]: Can you rephrase this as follows?

"Specifies the resources role that Jenkins scheduler will receive offers for. If role is set to "*" (default), it will receive offers for unreserved resources. If role is set to something else (e.g., 'jenkins'), it will receive offers for BOTH unreserved resources and 'jenkins' resources."

Thinking about this a bit more, do we want the Jenkins scheduler to receive offers for both unreserved and reserved roles? If only the latter, we will need some logic in JenkinsSheduler to reject unreserved resources.

**Comment by jduan** [comment]
Not sure why there was a failing test on CI. I've run "mvn test" many times on my laptop, and they passed every time.

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/help-role.html [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<div>
+    Specifies the resources role that Jenkins scheduler will receive offers
```
Comment (vinodkone) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/help-role.html]: As discussed can you rephrase this statement in terms what resources are used for launching tasks (and not in terms of what resources it receives)?

File: src/test/java/org/jenkinsci/plugins/mesos/JenkinsSchedulerTest.java [hunk 3]
```diff
@@ -249,7 +249,9 @@ public void testConstructMesosCommandInfoWithNullCustomDockerShell() throws Exce
                 null,               // externalContainerInfo,
                 containerInfo,      // containerInfo,
                 null);              //additionalURIs
-        Mesos.SlaveRequest slaveReq = new Mesos.SlaveRequest(new Mesos.JenkinsSlave(TEST_JENKINS_SLAVE_NAME),0.2d,TEST_JENKINS_SLAVE_MEM,mesosSlaveInfo);
+        Mesos.SlaveRequest slaveReq = new Mesos.SlaveRequest(
+                new Mesos.JenkinsSlave(TEST_JENKINS_SLAVE_NAME),0.2d,
```
Comment (vinodkone) [inline src/test/java/org/jenkinsci/plugins/mesos/JenkinsSchedulerTest.java]: can you put each argument on its own line? we do this when we have to wrap a statement.

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 4]
```diff
@@ -289,6 +290,13 @@ private boolean matches(Offer offer, Request request) {
     List<Range> ports = null;
 
     for (Resource resource : offer.getResourcesList()) {
+      String resourceRole = resource.getRole();
+      String expectedRole = mesosCloud.getRole();
+      if (! resourceRole.equals(expectedRole)) {
+        LOGGER.warning("Resource role " + resourceRole +
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: Don't think this should be a warning. s/warning/fine/

File: src/main/java/org/jenkinsci/plugins/mesos/Mesos.java [hunk 5]
```diff
@@ -35,12 +35,15 @@ public JenkinsSlave(String name) {
     JenkinsSlave slave;
     final double cpus;
     final int mem;
+    final String role;
     final MesosSlaveInfo slaveInfo;
 
-    public SlaveRequest(JenkinsSlave slave, double cpus, int mem, MesosSlaveInfo slaveInfo) {
+    public SlaveRequest(JenkinsSlave slave, double cpus, int mem, String role,
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/Mesos.java]: line wrapping. can you put one argument per line? what's the standard in java world?

File: src/main/java/org/jenkinsci/plugins/mesos/MesosComputerLauncher.java [hunk 6]
```diff
@@ -72,9 +72,10 @@ public void launch(SlaveComputer _computer, TaskListener listener) throws Interr
     // Create the request.
     double cpus = computer.getNode().getCpus();
     int mem = computer.getNode().getMem();
+    String role = cloud.getRole();
 
     Mesos.SlaveRequest request = new Mesos.SlaveRequest(new JenkinsSlave(name),
-        cpus, mem, computer.getNode().getSlaveInfo());
```
Comment (vinodkone) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosComputerLauncher.java]: wrapping.

**Comment by vinodkone** [comment]
Sorry for the delay in getting back. Looks good to me modulo minor comments.

Also, can you please squash the commits?

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 7]
```diff
@@ -289,6 +290,13 @@ private boolean matches(Offer offer, Request request) {
     List<Range> ports = null;
 
     for (Resource resource : offer.getResourcesList()) {
+      String resourceRole = resource.getRole();
+      String expectedRole = mesosCloud.getRole();
+      if (! resourceRole.equals(expectedRole)) {
```
Comment (felixb) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: the plugin should still accept `*` offers, even if set to a specific role.
marathon and chronos are doing the same.

this should be:

    if (! (resourceRole.equals(expectedRole) || resourceRole.equals("*"))) {

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 8]
```diff
@@ -518,6 +526,7 @@ private void detectAndAddExternalContainerInfo(Request request, CommandInfo.Buil
                 .newBuilder()
```
Comment (felixb) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: This is not working with offers of different roles, e.g. `cpus(jenkins):5;mem(*):10240`.
I changed the method to reflect this kind of situations:

      private TaskInfo.Builder getTaskInfoBuilder(Offer offer, Request request, TaskID taskId, CommandInfo.Builder commandBuilder) {
          TaskInfo.Builder builder = TaskInfo.newBuilder()
            .setName("task " + taskId.getValue())
            .setTaskId(taskId)
            .setSlaveId(offer.getSlaveId())
                  .setCommand(commandBuilder.build());
    
          double cpusNeeded = request.request.cpus;
          double memNeeded = request.request.mem;
    
          for (Resource r : offer.getResourcesList()) {
              if (r.getName().equals("cpus") && cpusNeeded > 0) {
                  double cpus = Math.min(r.getScalar().getValue(), cpusNeeded);
                  builder.addResources(
                          Resource
                                  .newBuilder()
                                  .setName("cpus")
                                  .setType(Value.Type.SCALAR)
                                  .setRole(r.getRole())
                                  .setScalar(
                                          Value.Scalar.newBuilder()
                                                  .setValue(cpus).build()).build());
                  cpusNeeded -= cpus;
              } else if (r.getName().equals("mem") && memNeeded > 0) {
                  double mem = Math.min(r.getScalar().getValue(),memNeeded);
                  builder.addResources(
                          Resource
                                  .newBuilder()
                                  .setName("mem")
                                  .setType(Value.Type.SCALAR)
                                  .setRole(r.getRole())
                                  .setScalar(
                                          Value.Scalar.newBuilder()
                                                  .setValue(mem).build()).build());
                  memNeeded -= mem;
              } else if (cpusNeeded == 0 && memNeeded == 0) {
                  break;
              }
          }
          return builder;
      }

**Comment by felixb** [comment]
@jduan I created a new PR #160 with a fix for this change. Feel squash it into your changes.

**Comment by Vlatombe** [comment]
Closing this PR in favor of #160
